src/gfn/gym/graph_building.py

```diff
@@ -0,0 +1,258 @@
+                add_node_out = torch.all(equal_nodes_per_batch == 0)
+
+        add_edge_mask = actions.action_type == GraphActionType.ADD_EDGE
+        if not torch.any(add_edge_mask):
+            add_edge_out = True
+        else:
+            add_edge_states = states[add_edge_mask].tensor
+            add_edge_actions = actions[add_edge_mask]
+            if torch.any(
+                add_edge_actions.edge_index[:, 0] == add_edge_actions.edge_index[:, 1]
+            ):
+                return False
+            if add_edge_states["node_feature"].shape[0] == 0:
+                return False
+            if torch.any(
+                add_edge_actions.edge_index > add_edge_states["node_feature"].shape[0]
+            ):
+                return False
+            global_edge_index = (
+                add_edge_actions.edge_index + add_edge_states["batch_ptr"][:-1][:, None]
+            )
+            equal_edges_per_batch = torch.all(
+                add_edge_states["edge_index"] == global_edge_index[:, None], dim=-1
+            ).sum(dim=-1)
+
+            if backward:
+                add_edge_out = torch.all(equal_edges_per_batch == 1)
+            else:
+                add_edge_out = torch.all(equal_edges_per_batch == 0)
+
+        return bool(add_node_out) and bool(add_edge_out)
+
+    def _add_node(
+        self,
+        tensor_dict: TensorDict,
+        batch_indices: torch.Tensor,
+        nodes_to_add: torch.Tensor,
+    ) -> TensorDict:
+        if isinstance(batch_indices, list):
+            batch_indices = torch.tensor(batch_indices)
+        if len(batch_indices) != len(nodes_to_add):
+            raise ValueError(
+                "Number of batch indices must match number of node feature lists"
+            )
+
+        modified_dict = tensor_dict.clone()
+        node_feature_dim = modified_dict["node_feature"].shape[1]
+
+        for graph_idx, new_nodes in zip(batch_indices, nodes_to_add):
+            tensor_dict["batch_ptr"][graph_idx]
+            end_ptr = tensor_dict["batch_ptr"][graph_idx + 1]
+
+            if new_nodes.ndim == 1:
+                new_nodes = new_nodes.unsqueeze(0)
+            if new_nodes.shape[1] != node_feature_dim:
+                raise ValueError(
+                    f"Node features must have dimension {node_feature_dim}"
+                )
+
+            # Update batch pointers for subsequent graphs
+            shift = new_nodes.shape[0]
+            modified_dict["batch_ptr"][graph_idx + 1 :] += shift
+
+            # Expand node features
+            modified_dict["node_feature"] = torch.cat(
+                [
+                    modified_dict["node_feature"][:end_ptr],
+                    new_nodes,
+                    modified_dict["node_feature"][end_ptr:],
+                ]
+            )
+
+            # Update edge indices
+            # Increment indices for edges after the current graph
+            edge_mask_0 = modified_dict["edge_index"][:, 0] >= end_ptr
```

Comment:
What does "after the current graph" mean here?

Reply:
Graphs states are saved in batches, i.e.
state = {
'node_features': (N, F)
'edge_indices': (M, 2)
batch_ptr: (B + 1,)
...
}

represent B graphs. When you add a node to graph `i` you must be sure to increase the edge_indices for every node in graphs after the batch `i`, so they represent the correct edges.

Reply:
Thanks -- I think perhaps more comments in the code would be helpful to ensure everyone is on the same page as to what is going on here (including future contributors).